### PR TITLE
Student view results: Fix anonymity breach for self-feedback questions #8203

### LINKED ISSUE #8203: Student view results: Fix anonymity breach for self-feedback questions
Question setting:
image

What students see:
image

If there is another anonymous question, students can figure out who gave each responses (because we use the same anonymous ID for all responses from the same student)

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #8203

**PR Checklist**

Ensure that you have:
- [x] Read and understood our PR guideline: https://github.com/TEAMMATES/teammates/blob/master/docs/process.md#step-4-submit-a-pr
  - [x] Added the issue number to the "Fixes" keyword above
  - [x] Titled the PR as specified in the abovementioned document
- [x] Made your changes on a branch other than `master` and `release`
- [x] Gone through all the changes in this PR and ensured that:
  - [x] They addressed one (and only one) issue
  - [x] No unintended changes were made
- [x] Run and passed static analysis: `./gradlew lint` and `npm run lint`
- [x] Added/updated tests, if changes in functionality were involved
- [x] Added/updated documentation to public APIs (classes, methods, variables), if applicable

**Outline of Solution**

Demo for self-feedback question.
image

Code to hide the random identifier for recipients is removed (refer to #5228) as it doesn't help according to the discussion in #8203. The hot fix applied to the LIVE server already achieves fully random identifiers and thus it is OK to show the random identifier for recipients.

### CODE REVIEW

**Comment by wkurniawan07** [comment]
@xpdavid there are still some failing tests there

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 1]
```diff
@@ -791,6 +791,10 @@ private void hideResponsesGiverRecipient() {
             String name = emailNameTable.get(response.recipient);
             FeedbackQuestionAttributes question = questions.get(response.feedbackQuestionId);
             FeedbackParticipantType participantType = question.recipientType;
+            if (participantType == FeedbackParticipantType.SELF) {
+                // recipient type for self-feedback is the same as the giver type
+                participantType = question.giverType;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Here

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 2]
```diff
@@ -836,10 +840,9 @@ public boolean isFeedbackParticipantVisible(boolean isGiver, FeedbackResponseAtt
             isVisible = visibilityTable.get(responseId)[Const.VISIBILITY_TABLE_RECIPIENT];
             participantType = question.recipientType;
         }
-        boolean isTypeSelf = participantType == FeedbackParticipantType.SELF;
         boolean isTypeNone = participantType == FeedbackParticipantType.NONE;
 
-        return isVisible || isTypeSelf || isTypeNone;
+        return isVisible || isTypeNone;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Here

File: src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java [hunk 3]
```diff
@@ -186,11 +186,6 @@ private FeedbackResultsResponseTable createResponseTable(FeedbackQuestionAttribu
                 // If the giver is the user, show the real name of the recipient
                 // since the giver would know which recipient he/she gave the response to
                 recipientName = bundle.getNameForEmail(response.recipient);
-            } else if (!isUserGiver
-                       && !bundle.isRecipientVisible(response)) {
-                // Hide anonymous recipient entirely to prevent student from guessing the identity
-                // based on responses from other response givers
-                recipientName = bundle.getAnonNameWithoutNumericalId(question.recipientType);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Here

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java (hunk 1 above)
Comment (xpdavid) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: For self-feedback question, `recipientType ` is always `FeedbackParticipantType.SELF` and therefore we need to get the name of the participant from `giverType`.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java (hunk 2 above)
Comment (xpdavid) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: This is the bug described in #8203. Before the change, self feedback question would always be visible regardless of the settings.

File: src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java (hunk 3 above)
Comment (xpdavid) [reply src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Refer to the discussion in #8203, there is no need to hide the random identifier for recipient anymore.

Comment (wkurniawan07) [reply src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Check if the method `getAnonNameWithoutNumericalId` can be removed.

**Pull request merged**